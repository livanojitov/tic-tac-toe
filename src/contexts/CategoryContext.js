import React, { createContext } from 'react';
import useFetchData from '../hooks/useFetchData';

export const CategoryContext = createContext();

const CategoryContextProvider = ({children}) => {
  let {data: categories, error} = useFetchData("categories");

  return (
    <CategoryContext.Provider value={{categories, error}}>
      {children}
    </CategoryContext.Provider>
  );

}
 
export default CategoryContextProvider;