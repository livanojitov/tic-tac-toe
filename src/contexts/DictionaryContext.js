import React, { createContext, useState, useEffect } from 'react';
import axios from 'axios'

export const DictionaryContext = createContext();

const DictionaryContextProvider = (props) => {
  const [DICTIONARY, setDictionary] = useState('');

  useEffect(() => {
    if (DICTIONARY.length === 0){
      axios.get('dictionary.json')
        .then(res => {
          setDictionary(res.data);
        })
    }
  });

  return (
    <DictionaryContext.Provider value={{ DICTIONARY }}>
      {props.children}
    </DictionaryContext.Provider>
  );

}
 
export default DictionaryContextProvider;