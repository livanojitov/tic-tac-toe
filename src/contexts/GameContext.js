import React, { createContext } from 'react';
import useLocalStorage          from '../hooks/useLocalStorage';
export const GameContext = createContext();

const GameContextProvider = ({children}) => {

  const {games, dispatch} = useLocalStorage('tictactoe', []);

  const getGame = (id) => {
    if (games.length > 0){
      return games.filter((game) => game.id === id)[0];
    } else {
      return false;
    }
  }

  return (
    <GameContext.Provider value={{games, dispatch, getGame}}>
      {children}
    </GameContext.Provider>
  );

}
 
export default GameContextProvider;