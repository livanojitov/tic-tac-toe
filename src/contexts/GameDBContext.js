import React, { createContext } from "react";
import useLocalStorage from "../hooks/useLocalStorage";
export const GameDBContext = createContext();

const GameDBContextProvider = ({ children }) => {
  const { games, dispatch } = useLocalStorage("tictactoe", []);

  const getGame = (id) => {
    if (games.length > 0) {
      return games.filter((game) => game.id === id)[0];
    } else {
      return false;
    }
  };

  return (
    <GameDBContext.Provider value={{ games, dispatch, getGame }}>
      {children}
    </GameDBContext.Provider>
  );
};

export default GameDBContextProvider;
