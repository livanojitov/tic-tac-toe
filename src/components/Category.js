import React, { Component } from 'react';
import { CategoryContext }  from '../contexts/CategoryContext';
import { HistoryContext }   from '../contexts/HistoryContext';
import * as constants from './Constants';

class Category extends Component{ 
  static contextType = CategoryContext;

  constructor(props){
    super(props);

    let categoryId = this.props.categoryId;
    if (!(typeof(categoryId) != undefined && categoryId >=0 && categoryId < 8)){
      categoryId = Math.floor(Math.random() * 8);
    }
 
    this.state = {
      categoryId
    }

  }
  
  render(){
    const { categories } = this.context;
    const { categoryId } = this.state;
    const { name } = categories[categoryId];
    let select;

    return (
      <HistoryContext.Consumer>{(historyContext) => {
      const { history } = historyContext;
      if (!history){  
        select =  (
          <select value={categoryId} onChange={this.onChange}>
            {categories.map(
              (category, index) => {
                return (
                  <option key={index} value={index}>{category.name}</option>  
                )
              })}
          </select>);
      }else{
        select = <span>{name}</span>;
      }

      return (
        <div className = "categories">
          <span>{constants.CATEGORY}: </span>
          {select}
        </div>
      )
    }}</HistoryContext.Consumer>
    )
  }

  componentDidUpdate(prevProps, prevState){
    if (prevState.categoryId !== this.state.categoryId){
      this.emitCategory();
    }     
  }

  shouldComponentUpdate(nextProps, nextState){
   return (this.state.categoryId !== nextState.categoryId);
  }

  componentDidMount(){
    this.emitCategory();    
  }

  emitCategory(){
    if (this.props.onCategoryChange){
      const categoryId = this.state.categoryId;
      this.props.onCategoryChange({
        categoryId
      });
    }
  }

  onChange = (e) => {
    const categoryId = e.target.value;
    this.setState(() => ({categoryId}));
  }


}

export default Category;
