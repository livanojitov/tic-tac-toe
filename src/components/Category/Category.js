import React, { useContext } from 'react';
import { CategoryContext }   from '../../contexts/CategoryContext';
import { DictionaryContext } from '../../contexts/DictionaryContext';
import './Category.css';
import PropTypes             from 'prop-types';

const Category = React.memo(({onChangeCategory = () => {}}) => {
  const { categories } = useContext(CategoryContext);  
  const { dictionary } = useContext(DictionaryContext);

  return dictionary && categories ? (
    <div className="categories" data-testid="categories">
      <span data-testid="description">{dictionary.CATEGORY}: </span>
      <select data-testid="select" onChange={(e) => {onChangeCategory(categories[e.target.value])}}>
      {categories.map((category) => {
          return (
            <option key={category.id} value={category.id} data-testid='select-option'>{dictionary[category.name]}</option>
          )
      })}
      </select>
    </div>
  ) : (null)
})

Category.propTypes = {
  onChangeCategory: PropTypes.func
}

export default Category;
