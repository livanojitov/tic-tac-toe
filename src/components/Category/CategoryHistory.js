import React, { useContext } from 'react';
import { CategoryContext }   from '../../contexts/CategoryContext';
import { DictionaryContext } from '../../contexts/DictionaryContext';
import './Category.css';

const CategoryHistory = ({categoryId}) => {
  const { categories } = useContext(CategoryContext);
  const { dictionary } = useContext(DictionaryContext);

  return dictionary && categories ? (
    <div className = "categories">
      <span>{dictionary.CATEGORY}: </span>
      <span>{dictionary[categories[categoryId].name]}</span>
    </div>
  ) : (null)
}

export default CategoryHistory
