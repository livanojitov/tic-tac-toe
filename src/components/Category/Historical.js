import React, { useContext } from "react";
import { CategoryContext } from "../../contexts/CategoryContext";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import "./Category.scss";

const Historical = ({ categoryId }) => {
  const { categories } = useContext(CategoryContext);
  const { dictionary } = useContext(DictionaryContext);

  return dictionary && categories ? (
    <div className="categories">
      <span>{dictionary.CATEGORY}: </span>
      <span>{dictionary[categories[categoryId].name]}</span>
    </div>
  ) : null;
};

export default Historical;
