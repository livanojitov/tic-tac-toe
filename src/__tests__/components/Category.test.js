import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { CategoryContext }   from '../../contexts/CategoryContext';
import { DictionaryContext } from '../../contexts/DictionaryContext';
import Category from '../../components/Category/Category';

describe('Category component', () => {
  let dictionary;
  let categories;
  let changeCategoryHandler;

  beforeEach(() => {
    dictionary = {ANIMALS:               "Animals",
                  FRUITS_AND_VEGETABLES: "Fruits & Vegetables", 
                  GERMANY_LANDMARKS:     "Germany landmarks", 
                  HAVANA_LANDMARKS:      "Havana landmarks", 
                  MONTREAL_LANDMARKS:    "Montreal landmarks", 
                  PUPPIES:               "Puppies", 
                  SEINFELD:              "Seinfeld", 
                  SOCCER_PLAYERS:        "Soccer players",
                  SPAIN_LANDMARKS:       "Spain landmarks",
                  VANCOUVER_LANDMARKS:   "Vancouver landmarks",
                  CATEGORY:              "Category"};

    categories = [{ id: 0, name: "HAVANA_LANDMARKS",      folder: "habana"    },
                  { id: 1, name: "MONTREAL_LANDMARKS",    folder: "montreal"  },
                  { id: 2, name: "VANCOUVER_LANDMARKS",   folder: "vancouver" },
                  { id: 3, name: "SPAIN_LANDMARKS",       folder: "spain"     },
                  { id: 4, name: "GERMANY_LANDMARKS",     folder: "germany"   },
                  { id: 5, name: "SOCCER_PLAYERS",        folder: "soccer"    },
                  { id: 6, name: "FRUITS_AND_VEGETABLES", folder: "fruits"    },
                  { id: 7, name: "ANIMALS",               folder: "animals"   },
                  { id: 8, name: "PUPPIES",               folder: "puppies"   },
                  { id: 9, name: "SEINFELD",              folder: "seinfeld"  }];

    changeCategoryHandler = jest.fn();
  });

  afterEach(cleanup);

  const renderCategory = () => {
    return render(
      <DictionaryContext.Provider value={{dictionary}}>
        <CategoryContext.Provider value={{categories}}>
          <Category onChangeCategory={changeCategoryHandler} />
        </CategoryContext.Provider>
      </DictionaryContext.Provider>
    );
  }

  test("renders the Category component correctly", () => {
    const { asFragment } = renderCategory();
    expect(asFragment()).toMatchSnapshot();
  })

  test("dictionary and category are not fetched yet", () => {
    dictionary = '';
    categories = '';
    const { queryByTestId } = renderCategory();
    expect(queryByTestId('categories')).toBeNull();
  })

  test("dictionary is not fetched yet but category is", () => {
    dictionary = '';
    const { queryByTestId } = renderCategory();
    expect(queryByTestId('categories')).toBeNull();
  })

  test("category is not fetched yet but dictionary is", () => {
    categories = '';
    const { queryByTestId } = renderCategory();
    expect(queryByTestId('categories')).toBeNull();
  })

  test("categories class", () => {
    const { getByTestId } = renderCategory();
    expect(getByTestId('categories')).toHaveClass('categories');
  })

  test("category description", () => {
    const { getByTestId } = renderCategory();
    expect(getByTestId('description')).toHaveTextContent(dictionary['CATEGORY']);
  })
  
  test("simulates selection", () => {
    const { getByTestId, getAllByTestId } = renderCategory();
    fireEvent.change(getByTestId('select'), { target: { value: 2 } });
    let options = getAllByTestId('select-option');
    
    expect(options[0].selected).toBeFalsy();
    expect(options[1].selected).toBeFalsy();
    expect(options[2].selected).toBeTruthy();
    expect(options[3].selected).toBeFalsy();
    expect(options[4].selected).toBeFalsy();
    expect(options[5].selected).toBeFalsy();
    expect(options[6].selected).toBeFalsy();
    expect(options[7].selected).toBeFalsy();
    expect(options[8].selected).toBeFalsy();
    expect(options[9].selected).toBeFalsy();
  })

  test("should change the current category", () => {
    const { getByTestId } = renderCategory();
    fireEvent.change(getByTestId('select'), { target: { value: 2 } });

    expect(changeCategoryHandler).toHaveBeenCalledTimes(1);
    expect(changeCategoryHandler).toHaveBeenCalledWith(categories[2]);
  });  
});
