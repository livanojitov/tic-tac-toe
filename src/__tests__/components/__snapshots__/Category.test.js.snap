// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category component renders the Category component correctly 1`] = `
<DocumentFragment>
  <div
    class="categories"
    data-testid="categories"
  >
    <span
      data-testid="description"
    >
      Category: 
    </span>
    <select
      data-testid="select"
    >
      <option
        data-testid="select-option"
        value="0"
      >
        Havana landmarks
      </option>
      <option
        data-testid="select-option"
        value="1"
      >
        Montreal landmarks
      </option>
      <option
        data-testid="select-option"
        value="2"
      >
        Vancouver landmarks
      </option>
      <option
        data-testid="select-option"
        value="3"
      >
        Spain landmarks
      </option>
      <option
        data-testid="select-option"
        value="4"
      >
        Germany landmarks
      </option>
      <option
        data-testid="select-option"
        value="5"
      >
        Soccer players
      </option>
      <option
        data-testid="select-option"
        value="6"
      >
        Fruits & Vegetables
      </option>
      <option
        data-testid="select-option"
        value="7"
      >
        Animals
      </option>
      <option
        data-testid="select-option"
        value="8"
      >
        Puppies
      </option>
      <option
        data-testid="select-option"
        value="9"
      >
        Seinfeld
      </option>
    </select>
  </div>
</DocumentFragment>
`;
